File: src/examples/ex_table_scene.cpp
/* ex_table_scene.cpp  (MJCF)
 * Kinova GEN3 + Robotiq 2F-85 gripper on a table with a few free objects,
 * loaded from MuJoCo Menagerie MJCF.
 *
 * The arm runs KDL gravity compensation; the gripper cycles open/closed every 3 s.
 *
 * Requires MuJoCo Menagerie in cache.
 *
 * Usage:
 *   ex_table_scene_mjcf [--headless]
 *
 * Runs 500 steps and exits; --headless skips the viewer and prints the final EE drift. */

#include "mj_kdl_wrapper/mj_kdl_wrapper.hpp"
#include "common.hpp"
#include "example_paths.hpp"

#include <kdl/chaindynparam.hpp>
#include <kdl/chainfksolverpos_recursive.hpp>

#include <cmath>
#include <iomanip>
#include <iostream>
#include <string>

using mj_kdl_examples::kHomePose;
static constexpr double kObjectMass  = 0.1;                    // [kg]
static constexpr double kFriction[3] = { 1.0, 0.005, 0.0001 }; // MuJoCo's geom default
static constexpr int    kSteps       = 500;

static mj_kdl::SceneObject make_box(
  const char *name,
  double      x,
  double      y,
  double      hx,
  double      hy,
  double      hz,
  float       r,
  float       g,
  float       b,
  double      surface_z
)
{
    mj_kdl::SceneObject o;
    o.name    = name;
    o.shape   = mj_kdl::Shape::BOX;
    o.size[0] = hx;
    o.size[1] = hy;
    o.size[2] = hz;
    o.pos[0]  = x;
    o.pos[1]  = y;
    o.pos[2]  = surface_z + hz;
    o.rgba[0] = r;
    o.rgba[1] = g;
    o.rgba[2] = b;
    o.rgba[3] = 1.0f;
    o.mass    = kObjectMass;
    for (int k = 0; k < 3; ++k) o.friction[k] = kFriction[k];
    return o;
}

static mj_kdl::SceneObject make_sphere(
  const char *name,
  double      x,
  double      y,
  double      radius,
  float       r,
  float       g,
  float       b,
  double      surface_z
)
{
    mj_kdl::SceneObject o;
    o.name    = name;
    o.shape   = mj_kdl::Shape::SPHERE;
    o.size[0] = radius;
    o.pos[0]  = x;
    o.pos[1]  = y;
    o.pos[2]  = surface_z + radius;
    o.rgba[0] = r;
    o.rgba[1] = g;
    o.rgba[2] = b;
    o.rgba[3] = 1.0f;
    o.mass    = kObjectMass;
    for (int k = 0; k < 3; ++k) o.friction[k] = kFriction[k];
    return o;
}

int main(int argc, char *argv[])
{
    const bool headless = mj_kdl_examples::parse_args(argc, argv).headless;

    const std::string mjcf     = mj_kdl_examples::menagerie_model("kinova_gen3/gen3.xml");
    const std::string grp_mjcf = mj_kdl_examples::asset("robotiq_2f85/2f85.xml");
    const std::string table_mjcf = mj_kdl_examples::asset("table.xml");

    mj_kdl::SceneSpec sc;
    sc.timestep   = 0.002;
    sc.add_floor  = true;
    sc.add_skybox = true;
    const double surface_z = 0.7;

    mj_kdl::AttachmentSpec gs;
    gs.mjcf_path = grp_mjcf;
    gs.attach_to = { mj_kdl::AttachKind::Site, "pinch_site" };
    gs.prefix    = "g_";

    mj_kdl::RobotSpec r;
    r.path   = mjcf;
    r.pos[2] = surface_z;
    r.attachments.push_back(gs);
    sc.robots.push_back(r);

    mj_kdl::SceneObject table{
        .name      = "table",
        .mjcf_path = table_mjcf,
        .pos       = { 0.0, 0.0, surface_z },
        .fixed     = true,
    };
    sc.objects.push_back(table);
    sc.objects.push_back(
      make_box("red_cube", 0.35, 0.10, 0.03, 0.03, 0.03, 1.0f, 0.2f, 0.2f, surface_z)
    );
    sc.objects.push_back(
      make_box("green_cube", 0.35, -0.10, 0.03, 0.03, 0.03, 0.2f, 1.0f, 0.2f, surface_z)
    );
    sc.objects.push_back(
      make_box("blue_cube", 0.35, 0.30, 0.04, 0.04, 0.04, 0.2f, 0.2f, 1.0f, surface_z)
    );
    sc.objects.push_back(
      make_sphere("orange_sphere", -0.20, 0.20, 0.035, 1.0f, 0.55f, 0.0f, surface_z)
    );
    sc.objects.push_back(
      make_sphere("purple_sphere", -0.20, -0.20, 0.025, 0.7f, 0.0f, 0.9f, surface_z)
    );

    /* Static scene cameras.  The Kinova MJCF also contributes a "wrist" camera;
     * all of them are in env.model after init_env(). */
    sc.cameras.push_back(mj_kdl::CameraSpec{
        .name = "overview",
        .pos  = { 0.0, -0.6, 1.6 }, // in front of and above the table
        // roll 34 deg: tilt down toward table
        .quat = { 0.29237170472273677, 0.0, 0.0, 0.9563047559630354 },
        .fovy = 55.0,
    });
    sc.cameras.push_back(mj_kdl::CameraSpec{
        .name = "side",
        .pos  = { -1.0, 0.0, 1.1 }, // left side, arm height
        // pitch -68 deg: pitch down toward robot base
        .quat = { 0.0, -0.5591929034707469, 0.0, 0.8290375725550416 },
        .fovy = 50.0,
    });

    mj_kdl::Env   env;
    mj_kdl::Robot robot;
    if (!mj_kdl::init_env(&env, &sc)) {
        std::cerr << "init_env() failed\n";
        return 1;
    }

    KDL::Frame world_T_table_top;
    if (!mj_kdl::get_site_frame(&env, "table_top", &world_T_table_top)) {
        std::cerr << "table_top site not found\n";
        return 1;
    }
    std::cout << "table top z = " << world_T_table_top.p.z() << "\n";

    std::cout << "cameras:";
    for (int i = 0; i < env.model->ncam; ++i)
        std::cout << " " << mj_id2name(env.model, mjOBJ_CAMERA, i);
    std::cout << "\n";

    const mj_kdl::ToolFrameSpec tool{ .tool_body = "g_base_mount", .tcp_site = "g_pinch" };

    if (!mj_kdl::init_robot_from_mjcf(&robot, &env, "base_link", "bracelet_link", "", &tool)) {
        std::cerr << "init_robot_from_mjcf() failed\n";
        return 1;
    }

    unsigned                        n = static_cast<unsigned>(robot.n_joints);
    KDL::ChainFkSolverPos_recursive fk(robot.chain);
    KDL::ChainDynParam              dyn(robot.chain, KDL::Vector(0.0, 0.0, sc.gravity_z));

    KDL::JntArray q_home(n);
    for (unsigned i = 0; i < n; ++i) q_home(i) = kHomePose[i];

    mj_kdl::SceneActuatorSlot *fingers = mj_kdl::bind_scene_actuator(&env.scene, "g_fingers_actuator");
    if (!fingers) return 1;

    if (!mj_kdl::set_control_mode(&robot, mj_kdl::CtrlMode::TORQUE)) return 1;

    // reset() seeds the finger slot from ctrl, so the hook sets ctrl.
    env.on_reset = [&](mj_kdl::ResetContext *ctx) {
        mj_kdl::set_joint_pos(&robot, q_home);
        ctx->data->ctrl[fingers->ctrl_id] = mj_kdl_examples::kGripperClosed;
    };

    mj_kdl::reset(&env);

    KDL::JntArray q(n), g(n);
    auto          ctrl_step = [&]() {
        mj_kdl::update(&env);
        for (unsigned i = 0; i < n; ++i) q(i) = robot.jnt_pos_msr[i];
        dyn.JntToGravity(q, g);
        for (unsigned i = 0; i < n; ++i) robot.jnt_trq_cmd[i] = g(i);
        fingers->command =
          (std::fmod(env.data->time, 6.0) < 3.0) ? mj_kdl_examples::kGripperClosed : 0.0;
    };

    if (headless) {
        KDL::Frame ee_start;
        fk.JntToCart(q_home, ee_start);

        for (int step = 0; step < kSteps; ++step) {
            ctrl_step();
            mj_kdl::step(&env);
        }

        KDL::JntArray q_end(n);
        for (unsigned i = 0; i < n; ++i) q_end(i) = robot.jnt_pos_msr[i];
        KDL::Frame ee_end;
        fk.JntToCart(q_end, ee_end);
        double drift = (ee_start.p - ee_end.p).Norm();
        std::cout << "EE drift after " << kSteps << " steps: " << std::fixed << std::setprecision(3)
                  << drift * 1000.0 << " mm\n";
    } else {
        if (!mj_kdl::open_viewer(&env)) {
            std::cerr << "open_viewer() failed\n";
            return 1;
        }
        for (int step = 0; step < kSteps; ++step) {
            ctrl_step();
            if (!mj_kdl::step(&env)) break;
            mj_kdl::pace_realtime(&env);
        }
    }

    mj_kdl::cleanup(&env);
    return 0;
}

File: python/examples/ex_table_scene.py
#!/usr/bin/env python3
"""Table-scene example ported from src/examples/ex_table_scene.cpp.

Arm + gripper on a table with free objects. The arm runs KDL gravity
compensation and the gripper cycles open/closed. An Env on_reset hook re-homes
the arm so the simulate-UI reset button restores the scene.
"""

from __future__ import annotations

import argparse
import math
import os
from pathlib import Path

import mj_kdl_wrapper as mjk

SURFACE_Z = 0.7
DEFAULT_MODEL = "third_party/menagerie/kinova_gen3/gen3.xml"
DEFAULT_GRIPPER = "assets/robotiq_2f85/2f85.xml"
DEFAULT_TABLE = "assets/table.xml"
MODEL_ENV_VAR = "MJ_KDL_MODEL"
GRIPPER_ENV_VAR = "MJ_KDL_GRIPPER"
TABLE_ENV_VAR = "MJ_KDL_TABLE"
HOME_POSE = [0.0, 0.2618, 3.1416, -2.2689, 0.0, 0.9599, 1.5708]


def path_from_arg(value: str, label: str) -> Path:
    path = Path(value)
    if not path.exists():
        raise FileNotFoundError(f"{value} does not exist for {label}")
    return path


def attachment_gripper(path: Path) -> mjk.AttachmentSpec:
    spec = mjk.AttachmentSpec()
    spec.mjcf_path = str(path)
    spec.attach_to = mjk.AttachTarget(mjk.AttachKind.Site, "pinch_site")
    spec.prefix = "g_"
    return spec


def table_object(path: Path) -> mjk.SceneObject:
    obj = mjk.SceneObject()
    obj.name = "table"
    obj.mjcf_path = str(path)
    obj.pos = [0.0, 0.0, SURFACE_Z]
    obj.fixed = True
    return obj


def make_box(name, x, y, hx, hy, hz, color) -> mjk.SceneObject:
    obj = mjk.SceneObject()
    obj.name = name
    obj.shape = mjk.Shape.BOX
    obj.size = [hx, hy, hz]
    obj.pos = [x, y, SURFACE_Z + hz]
    obj.rgba = [color[0], color[1], color[2], 1.0]
    obj.mass = 0.2
    obj.friction = [1.0, 0.005, 0.0001]
    return obj


def make_sphere(name, x, y, radius, color) -> mjk.SceneObject:
    obj = mjk.SceneObject()
    obj.name = name
    obj.shape = mjk.Shape.SPHERE
    obj.size = [radius, 0.0, 0.0]
    obj.pos = [x, y, SURFACE_Z + radius]
    obj.rgba = [color[0], color[1], color[2], 1.0]
    obj.mass = 0.1
    obj.friction = [1.0, 0.005, 0.0001]
    return obj


def scene_objects(table_path):
    return [
        table_object(table_path),
        make_box("red_cube", 0.35, 0.10, 0.03, 0.03, 0.03, (1.0, 0.2, 0.2)),
        make_box("green_cube", 0.35, -0.10, 0.03, 0.03, 0.03, (0.2, 1.0, 0.2)),
        make_box("blue_cube", 0.35, 0.30, 0.04, 0.04, 0.04, (0.2, 0.2, 1.0)),
        make_sphere("orange_sphere", -0.20, 0.20, 0.035, (1.0, 0.55, 0.0)),
        make_sphere("purple_sphere", -0.20, -0.20, 0.025, (0.7, 0.0, 0.9)),
    ]


def build_env(model_path: Path, gripper_path: Path, table_path: Path) -> tuple[mjk.Env, mjk.Robot]:
    spec = mjk.SceneSpec()
    spec.timestep = 0.002
    spec.add_floor = True
    spec.add_skybox = True
    spec.objects = scene_objects(table_path)
    robot_spec = mjk.RobotSpec()
    robot_spec.path = str(model_path)
    robot_spec.pos = [0.0, 0.0, SURFACE_Z]
    robot_spec.attachments = [attachment_gripper(gripper_path)]
    spec.robots = [robot_spec]
    overview = mjk.CameraSpec()
    overview.name = "overview"
    overview.pos = [0.0, -0.6, 1.6]
    overview.euler = [34.0, 0.0, 0.0]
    overview.fovy = 45.0
    side = mjk.CameraSpec()
    side.name = "side"
    side.pos = [-1.0, 0.0, 1.1]
    side.euler = [0.0, -68.0, 0.0]
    side.fovy = 45.0
    spec.cameras = [overview, side]
    env = mjk.Env.build(spec)
    tool = mjk.ToolFrameSpec()
    tool.tool_body = "g_base"
    tool.tcp_site = "g_pinch"
    robot = env.create_robot("base_link", "bracelet_link", tool=tool)
    return env, robot


def run_loop(env: mjk.Env, robot: mjk.Robot, step_fn, *, duration: float, gui: bool) -> None:
    if gui:
        viewer = mjk.SimulateViewer.open(robot, "ex_table_scene.py")
        prev = env.time()
        try:
            while viewer.is_running():
                if env.time() < prev - 1e-6:
                    env.reset()
                prev = env.time()
                step_fn()
                if not viewer.step():
                    break
        finally:
            viewer.close()
        return
    end = env.time() + duration
    while env.time() < end:
        step_fn()
        robot.step()


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("--gui", action="store_true")
    args = parser.parse_args()

    env, robot = build_env(
        path_from_arg(mjk.menagerie.model_path("kinova_gen3", env_var=MODEL_ENV_VAR), "arm model"),
        path_from_arg(os.environ.get(GRIPPER_ENV_VAR, DEFAULT_GRIPPER), "gripper model"),
        path_from_arg(os.environ.get(TABLE_ENV_VAR, DEFAULT_TABLE), "table model"),
    )
    try:
        robot.ctrl_mode = mjk.CtrlMode.TORQUE
        env.on_reset = lambda ctx: robot.set_joint_pos(HOME_POSE, call_forward=False)
        env.reset()

        top = env.site_frame(mjk.scene_object_site_name(env.spec.objects[0], "table_top"))
        print(f"table top z = {top.p.z():.3f}")
        print(f"cameras: {' '.join(env.camera_names())}")

        def step():
            robot.update()
            robot.jnt_trq_cmd = robot.gravity_torques(-9.81)
            if env.has_actuator("g_fingers_actuator"):
                env.set_actuator_ctrl(
                    "g_fingers_actuator", 255.0 if math.fmod(env.time(), 6.0) < 3.0 else 0.0
                )

        run_loop(env, robot, step, duration=1.0, gui=args.gui)
    finally:
        env.close()
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
